File: Template/Naked Objects Server/Template.Model/SetRepository.cs
using NakedObjects;
using System.Linq;

namespace Template.Model
{
    [Named("Sets")]
    public class SetRepository
    {
        #region Injected Services
        //An implementation of this interface is injected automatically by the framework
        public IDomainObjectContainer Container { set; protected get; }
        #endregion

        public Set CreateNewSet() => Container.NewTransientInstance<Set>();

        public IQueryable<Set> ListSets([Optionally] Subject subject, [Optionally] int? yearGroup)
        {
            var sets = Container.Instances<Set>();
            if (subject != null)
            {
                int id = subject.Id;
                sets = sets.Where(s => s.Subject.Id == id);
            }
            if (yearGroup != null)
            {
                sets = sets.Where(s => s.YearGroup == yearGroup.Value);
            }
            return sets.OrderBy(s => s.YearGroup).ThenBy(s => s.Subject.Name);
        }
    }
}

File: Template/NOF2 Server/Model.Template/SetRepository.vb
Public Class SetRepository
    Public Property Container As IContainer

    Public Function CreateNewSet() As [Set]
        Return Container.CreateTransientInstance(Of [Set])()
    End Function

    Public Function ListSets(
 ByVal subject As Subject,
 ByVal yearGroup As Integer?) As IQueryable(Of [Set])
        Dim sets = Container.AllInstances(Of [Set])()

        If subject IsNot Nothing Then
            Dim id As Integer = subject.Id
            sets = sets.Where(Function(s) s.Subject.Id = id)
        End If

        If yearGroup IsNot Nothing Then
            sets = sets.Where(Function(s) s.YearGroup = yearGroup.Value)
        End If

        Return sets.OrderBy(Function(s) s.YearGroup).ThenBy(Function(s) s.Subject.Name)
    End Function
End Class

